hw/ip/pattgen/doc/_index.md

```diff
@@ -0,0 +1,143 @@
+
+Each FSM consists of.
+- Inputs::
+    - `clk_io`, `reset`, `enable`, `clk_pre_divide`, `pattern`, `pattern_size`, `polarity`,  and `n_repeats`
+- Outputs:
+    - `pda` and `pcl`
+- a single state variable with three states `IDLE`, `ACTIVE`, and `END`,.
+- a clock-divide counter, `clk_div`. a single clock-divide flop, `clk_int`, and
+- two two internal counters `bit_ctr` and `repeat_ctr`
+
+Each FSM is disabled when `enable` is low.
+Disabling the FSM is equivalent to reset, and both operations place the FSM in the `IDLE` state.
+While in `IDLE`, the other state machine registers assume their default states:
+The internal counters, the clock-divide, `bit_ctr` and `repeat_ctr` all reset to 0, as does `clk_int`.
+
+Once the FSM is enabled, it transitions to the `ACTIVE` state.  The clock-divide counter `clk_div` increments every cycle, except when it overflows at `clk_pre_divide`.
+Then it resets to 0, toggling `clk_int` in the process.
+The results in an internal clock frequency of:
+$$f_\textsf{clk\_int}=\frac{f_\textsf{clk\_io}}{2\times\textsf{clkPreDivide}}\textrm{ for }x\in{0, 1}$$.
+
+The FSM clock output, `pcl`, is directly driven by `clk_int`, unless the `polarity` input is high, in which case `pcl` is inverted from `clk_int`.
+
+The `bit_ctr` counter increments on every falling edge of `clk_int`, until it overflows at the pattern length based on the `pattern_size` input.
+
+In the `ACTIVE` state, the FSM `pda` output is driven by a multiplexer, connected to the `pattern` input.
+The value of `bit_ctr` selects the bit value from the appropriate sequence position, via this multiplexor.
+
+Finally whevever `bit_ctr` overflows and reverts to zero, the `repeat_ctr` increments, and the pattern starts again.
+Finally `repeat_ctr` overflows to zero as it reaches the input value `n_repeats`.
+When this overflow occurs, the FSM transitions to the `END` state.
+All counters halt, the `pda` data lines reset to zero, and an interupt event is sent out to signal completion.
+
+The entire sequence can be restarted either by resetting or disabling and re-enabling the FSM.
+
+### Interrupts
+
+PATTGEN HWIP provides two interrupt pins, `patt_done0` and `patt_done1`, which 
+indicate the completion of pattern generation on the output channels. 
+These interrupts can be enabled/disabled by setting/un-setting the corresponding 
+bits of `intr_enable` register. To clear the interrupts, 
+bit `1` must be written the corresponding bits of `intr_state` register
+
+# Programmers guide
+
+## Initialization
+
+After reset and enabling, the initialization of the PATTGEN HWIP primarily consists of 4 steps:
+1. Disabling the PATTGEN functionality, for the desired channels,
+1. intializating the control registers,
+1. configuring the interrupts, and
+1. enabling the PATTGEN functionality for the desired channels.
```

Comment:
This information is effectively duplicated from the "Theory of Operations" section above. I'd probably move the description of the programming sequence from "Theory of Operations" here, and put a rather generic sentence like "The pattern generator is programmed with the desired pattern, its length, and an appropriate output clock configuration. After enabling it, the pattern is repeatedly output until the pattern generator is disabled again." into the "Theory of Operations" section at the start.

Reply:
I like that idea.  Thanks for the suggestion.